Ruleset: AsyncAPI (#974)
stoplightio/spectral: docs/reference/asyncapi-rules.md

> # AsyncAPI Rules

Reviewer: Is this file incomplete?

Author: @fmvilas Before diving into the documentation part of the PR, I was expecting some comments from @philsturgeon with regards to the ruleset first.

This file will eventually list all the available rules.

Author: @fmvilas @philsturgeon The file now list all the proposed rules. Could you please take a look at them?

Every re-wording proposal would be greatly appreciated! ;-)

# AsyncAPI Rules

Spectral has a built-in "asyncapi" ruleset for the [AsyncAPI Specification](https://www.asyncapi.com/docs/specifications/2.0.0/).

In your ruleset file you can add `extends: "spectral:asyncapi"` and you'll get all of the following rules applied.

> ## AsyncAPI v2-only

Reviewer 2: ```suggestion
```

we only had this in here when we were documenting both oas2 and oas3 at the same time.

Author: @philsturgeon Fixed

These rules will only apply to AsyncAPI v2 documents.

> ### asyncapi2-channel-no-empty-parameter

Reviewer 2: We only had version numbers in the rules when a rule only applied to a specific version. We'll never support `asyncapi1` and we dont know what 3 will look like, these rules might apply and we'd have to rename them later. I think we'd be better off naming them all `asyncapi-*` now to avoid breaking later.

Author: @philsturgeon Fixed

Channel parameter declarations cannot be empty, ex.`/given/{}` is invalid.

**Recommended:** Yes

### asyncapi2-channel-no-query-nor-fragment

> Query parameters and fragments shouldn't be used in channel names. Instead, use bindings to define them.

Reviewer 2: Can we get some examples?

Author: @philsturgeon I'm afraid I won't be able to come up with a good/bad example that would make sense in an AsyncAPI context.

@fmvilas Could you please help me here? What's the known bad practices regarding query and fragments and how should they be converted to bindings?

**Recommended:** Yes

### asyncapi2-channel-no-trailing-slash

> Keep trailing slashes off of channel names, as it can cause some confusion. Most messaging protocols will treat `example/foo` and `example/foo/` as different things. Keep in mind that tooling may replace slashes (`/`) with protocol-specific notation (e.g., `.` for AMQP), therefore, a trailing slash may result in an invalid channel name in some protocols.

Reviewer 2: ```suggestion
Keep trailing slashes off of channel names, as it can cause some confusion. Most messaging protocols will treat `example/foo` and `example/foo/` as different things. Keep in mind that tooling may replace slashes (`/`) with protocol-specific notation (e.g.: `.` for AMQP), therefore, a trailing slash may result in an invalid channel name in some protocols.
```

Author: @philsturgeon Fixed

Author: @philsturgeon Fixed

**Recommended:** Yes

### asyncapi2-headers-schema-type-object

The schema definition of the application headers must be of type “object”.

**Recommended:** Yes

### asyncapi2-info-contact-properties

> The [info-contact](#asyncapi2-info-contact) rule will ask you to put in a contact object, and this rule will make sure it's full of the most useful properties: `name`, `url` and `email`.

Reviewer 2: ```suggestion
The [asyncapi-info-contact](#asyncapi-info-contact) rule will ask you to put in a contact object, and this rule will make sure it's full of the most useful properties: `name`, `url` and `email`.
```
these should match

Author: @philsturgeon Fixed

Author: @philsturgeon Fixed

```yaml
asyncapi: "2.0.0"
info:
  license:
    name: MIT
    url: https://www.tldrlegal.com/l/mit
```

### asyncapi2-info-license

The `info` object should have a `license` key.

It can be hard to pick a license, so if you don't have a lawyer around you can use [TLDRLegal](https://tldrlegal.com/) and [Choose a License](https://choosealicense.com/) to help give you an idea.

How useful this is in court is not entirely known, but having a license is better than not having a license.

**Recommended:** Yes

**Good Example**

```yaml
asyncapi: "2.0.0"
info:
  license:
    name: MIT
```

### asyncapi2-operation-description

Operation objects should have a description.

**Recommended:** Yes

### asyncapi2-operation-operationId

This operation ID is essentially a reference for the operation. Tools may use it for defining function names, class method names, and even URL hashes in documentation systems.

**Recommended:** Yes

### asyncapi2-parameter-description

Parameter objects should have a `description`.

**Recommended:** No

### asyncapi2-payload-default

`default` objects should be valid against the `payload` they decorate.

**Recommended:** Yes

### asyncapi2-payload-examples

> Values of the `examples` array should be valid against the `payload` they decorate.

Reviewer 2: examples here and the default rule above too please, its super valuable and hard to explain.

Author: @philsturgeon Fixed

Author: @philsturgeon Could you please double check the proposed examples are clear enough?

```yaml
tags:
  - name: Invoice Items
    description: |+
      Giant long explanation about what this business concept is, because other people _might_ not have a clue!
```

**Recommended:** No

### asyncapi2-tags-alphabetical

AsyncAPI object should have alphabetical `tags`. This will be sorted by the `name` property.

**Recommended:** No

**Bad Example**

```yaml
tags:
  - name: 'Badger'
  - name: 'Aardvark'
```

**Good Example**

```yaml
tags:
  - name: 'Aardvark'
  - name: 'Badger'
```

**Recommended:** No

### asyncapi2-tags

AsyncAPI object should have non-empty `tags` array.

Why? Well, you _can_ reference tags arbitrarily in operations, and definition is optional...

```yaml
/invoices/{id}/items:
  get:
    tags:
    - Invoice Items
```

Defining tags allows you to add more information like a `description`. For more information see [tag-description](#asyncapi2-tag-description).

**Recommended:** Yes

### asyncapi2-unused-components-schema

Potential unused reusable `schema` entry has been detected.

> _Warning:_ This rule may identify false positives when linting a specification

Reviewer 2: ```suggestion
<!-- theme: warning -->
*Warning:* This rule may identify false positives when linting a specification
```

Author: @philsturgeon Fixed.

# AsyncAPI Rules

Spectral has a built-in "asyncapi" ruleset for the [AsyncAPI Specification](https://www.asyncapi.com/docs/specifications/2.0.0/).

In your ruleset file you can add `extends: "spectral:asyncapi"` and you'll get all of the following rules applied.

## AsyncAPI v2-only

These rules will only apply to AsyncAPI v2 documents.

### asyncapi2-channel-no-empty-parameter

> Path parameter declarations cannot be empty, ex.`/given/{}` is invalid.

Reviewer: ```suggestion
Channel parameter declarations cannot be empty, ex.`/given/{}` is invalid.
```

Author: @fmvilas Fixed

**Recommended:** Yes

### asyncapi2-channel-no-query-nor-fragment

> Query parameters and fragments shouldn't be used in paths. Instead use bindings to define them.

Reviewer: ```suggestion
Query parameters and fragments shouldn't be used in channel names. Instead, use bindings to define them.
```

Author: @fmvilas Fixed

**Recommended:** Yes

### asyncapi2-channel-no-trailing-slash

> Keep trailing slashes off of paths, as it can cause some confusion. Some web tooling (like mock servers, real servers, code generators, application frameworks, etc.) will treat `example.com/foo` and `example.com/foo/` as the same thing, but other tooling will not. Avoid any confusion by just documenting them without the slash, and maybe some tooling will let people shove a / on there when they're using it or maybe not, but at least the docs are suggesting how it should be done properly.

Reviewer: ```suggestion
Keep trailing slashes off of channel names, as it can cause some confusion. Most messaging protocols will treat `example/foo` and `example/foo/` as different things. Keep in mind that tooling may replace slashes (`/`) with protocol-specific notation (e.g., `.` for AMQP), therefore, a trailing slash may result in an invalid channel name in some protocols.
```

Author: @fmvilas Fixed

    ## Authentication

    This API uses OAuth2 and tokens can be requested from [Dev Portal: Tokens](https://example.org/developers/tokens).
```

### asyncapi2-info-license-url

Mentioning a license is only useful if people know what the license means, so add a link to the full text for those who need it.

**Recommended:** No

**Good Example**

```yaml
asyncapi: "2.0.0"
info:
  license:
    name: MIT
    url: https://www.tldrlegal.com/l/mit
```

### asyncapi2-info-license

The `info` object should have a `license` key.

It can be hard to pick a license, so if you don't have a lawyer around you can use [TLDRLegal](https://tldrlegal.com/) and [Choose a License](https://choosealicense.com/) to help give you an idea.

How useful this is in court is not entirely known, but having a license is better than not having a license.

**Recommended:** Yes

**Good Example**

```yaml
asyncapi: "2.0.0"
info:
  license:
    name: MIT
```

### asyncapi2-operation-description

Operation objects should have a description.

**Recommended:** Yes

### asyncapi2-operation-operationId

> This operation ID is essentially a reference for the operation, which can be used to visually suggest a connection to other operations. This is like some theoretical static HATEOAS-style referencing, but it's also used for the URL in some documentation systems.

Reviewer: I think this description is very http-ish. There's no HATEOAS or URLs in most of the messaging systems.

```suggestion
This operation ID is essentially a reference for the operation. Tools may use it for defining function names, class method names, and even URL hashes in documentation systems.
```

Author: @fmvilas Fixed

> Make the value `lower-hyphen-case`, and try and think of a name for the action which does not relate to the HTTP message. Base it off the actual action being performed. `create-polygon`? `search-by-polygon`? `filter-companies`?

Reviewer: Why this recommendation? I'd say that camelCase is the one that works the best in most places. Most of the times the operationId is used for function or method names and URL hashes. `lower-hyphen-case` is usually not a friendly notation in programming languages and URL hashes also support camelCase.

Author: @fmvilas Dropped.

**Recommended:** Yes

### asyncapi2-parameter-description

Parameter objects should have a `description`.

**Recommended:** No

### asyncapi2-payload-default

`default` objects should be valid against the `payload` they decorate.

**Recommended:** Yes

### asyncapi2-payload-example

> `example` objects should be valid against the `payload` they decorate.

Reviewer: This one doesn't exist on AsyncAPI. Only `examples`.

Author: @fmvilas Dropped.

**Recommended:** Yes

### asyncapi2-payload-examples

Values of the `examples` array should be valid against the `payload` they decorate.

**Recommended:** Yes

### asyncapi2-payload-unsupported-schemaFormat

Spectral currently only supports payload validation against the default AsyncAPI schema.

This rule will notify you when a payload is decorated with a `schemaFormat` property. Spectral won't be able to validate it and potential issues in this payload will go undetected by the `asyncapi2-payload-*` rules.

**Recommended:** Yes

### asyncapi2-payload

When `schemaFormat` is undefined, the `payload` object should be valid against the AsyncAPI 2 Schema Object definition.

**Recommended:** Yes

### asyncapi2-schema-default

`default` objects should be valid against the `schema` they decorate.

**Recommended:** Yes

### asyncapi2-schema-example

> `example` objects should be valid against the `schema` they decorate.

Reviewer: Same as above. This one doesn't exist on AsyncAPI.

Author: @fmvilas Dropped.

**Recommended:** Yes

### asyncapi2-schema-examples

Values of the `examples` array should be valid against the `schema` they decorate.

**Recommended:** Yes

### asyncapi2-schema

Validate structure of AsyncAPI v2 specification.

**Recommended:** Yes

### asyncapi2-server-no-empty-variable

Server URL variable declarations cannot be empty, ex.`gigantic-server.com/{}` is invalid.

**Recommended:** Yes

### asyncapi2-server-no-trailing-slash

Server URL should not have a trailing slash.

> Some tooling forgets to strip trailing slashes off when it's joining the `servers.url` with `channels`, and you can get awkward URLs like `https://example.com/api//pets`. Best to just strip them off yourself.

Reviewer: ```suggestion
Some tooling forgets to strip trailing slashes off when it's joining the `servers.url` with `channels`, and you can get awkward URLs like `https://example.com/api//pets`. Best to just strip them off yourself.
```

Author: @fmvilas Fixed (switched to mqtt protocol)

**Recommended:** Yes

**Good Example**

``` yaml
servers:

>   - url: https://example.com
>   - url: https://example.com/api

Reviewer: It would be great not to use HTTP as it's one of the less common protocols in messaging.

```suggestion
  - url: mqtt://example.com
  - url: mqtt://example.com/broker
```

Author: @fmvilas Fixed

```

**Bad Example**

``` yaml
servers:

>   - url: https://example.com/
>   - url: https://example.com/api/

Reviewer: ```suggestion
  - url: mqtt://example.com/
  - url: mqtt://example.com/broker/
```

Author: @fmvilas Fixed

Putting in the name of the developer/team/department/company responsible for the API, along with the support email and help-desk/GitHub Issues/whatever URL means people know where to go for help. This can mean more money in the bank, instead of developers just wandering off or complaining online.

**Recommended:** Yes

**Good Example**

```yaml
asyncapi: "2.0.0"
info:
  title: Awesome API
  description: A very well defined API
  version: "1.0"
  contact:
    name: A-Team
    email: a-team@goarmy.com
    url: https://goarmy.com/apis/support
```

### asyncapi-info-contact

Info object should contain `contact` object.

Hopefully your API description document is so good that nobody ever needs to contact you with questions, but that is rarely the case. The contact object has a few different options for contact details.

**Recommended:** Yes

**Good Example**

```yaml
asyncapi: "2.0.0"
info:
  title: Awesome API
  version: "1.0"
  contact:
    name: A-Team
    email: a-team@goarmy.com
```

### asyncapi-info-description

> OpenAPI object info `description` must be present and non-empty string.

Reviewer 2: Oops! 
```suggestion
AsyncAPI object info `description` must be present and non-empty string.
```

Author: @philsturgeon Fixed

Examples can contain Markdown so you can really go to town with them, implementing getting started information like where to find authentication keys, and how to use them.

**Recommended:** Yes

**Good Example**

```yaml
asyncapi: "2.0.0"
info:
  version: "1.0.0"
  title: Descriptive API
  description: >+
    Some description about the general point of this API, and why it exists when another similar but different API also exists.

>     ## Authentication
> 
>     This API uses OAuth2 and tokens can be requested from [Dev Portal: Tokens](https://example.org/developers/tokens).

Reviewer 2: ```suggestion
```

Author: @philsturgeon Fixed

**Recommended:** Yes

### asyncapi-parameter-description

Parameter objects should have a `description`.

**Recommended:** No

### asyncapi-payload-default

`default` objects should be valid against the `payload` they decorate.

**Recommended:** Yes

**Good Example**

``` yaml
payload:
  type: object
  properties:
    value:
      type: integer
  required:
    - value
  default:
    value: 17
```

**Bad Example**

``` yaml
payload:
  type: object
  properties:
    value:
      type: integer
  required:
    - value
  default:
    value: nope!
```

### asyncapi-payload-examples

Values of the `examples` array should be valid against the `payload` they decorate.

**Recommended:** Yes

**Good Example**

``` yaml
payload:
  type: object
  properties:
    value:
      type: integer
  required:
    - value
  examples:
    - value: 13
    - value: 17
```

**Bad Example**

``` yaml
payload:
  type: object
  properties:
    value:
      type: integer
  required:
    - value
  examples:
    - value: nope!       # Wrong type
    - notGoodEither: 17  # Missing required property
```

### asyncapi-payload-unsupported-schemaFormat

> Spectral currently only supports payload validation against the default AsyncAPI schema.
> 
> This rule will notify you when a payload is decorated with a `schemaFormat` property. Spectral won't be able to validate it and potential issues in this payload will go undetected by the `asyncapi-payload-*` rules.
> 

Reviewer 2: ```suggestion
AsyncAPI can support various `schemaFormat` values, with the default being one of the following:

```
application/vnd.aai.asyncapi;version=2.0.0
application/vnd.aai.asyncapi+json;version=2.0.0
application/vnd.aai.asyncapi+yaml;version=2.0.0
```

At this point no other schemaFormat's are supported by Spectral, so if you use them this rule will emit an `info` message and skip validating the payload. 

Other formats such as OpenAPI Schema Object, JSON Schema Draft 07 and Avro will be added in various upcoming versions. 
```

Author: @philsturgeon Awesome! I've slightly reworded your proposal (as we don't support setting the `schemaFormat` at all).
